src/core/ucc_lib.c

```diff
@@ -0,0 +1,220 @@
+    need_ccm_name_check = (0 != strcmp(config->ccms.names[0], "all"));
+    lib->n_libs_opened = 0;
+    for (i=0; i<n_ccms; i++) {
+        ccm_iface = ucc_lib_data.ccm_ifaces[i];
+        if (need_ccm_name_check &&
+            -1 == ucs_config_names_search(config->ccms, ccm_iface->name)) {
+            continue;
+        }
+        CHECK_LIB_CONFIG_CAP(thread_mode,  THREAD_MODE);
+        CHECK_LIB_CONFIG_CAP(coll_types,   COLL_TYPES);
+        ccm_config  = malloc(ccm_iface->ccm_lib_config.size);
+        ucs_config_parser_fill_opts(ccm_config, ccm_iface->ccm_lib_config.table,
+                                    config->full_prefix, ccm_iface->ccm_lib_config.prefix, 0);
+        status = ccm_iface->init(params, config, ccm_config, &ccm_lib);
+
+        if (UCS_OK != status) {
+            ucs_config_parser_release_opts(ccm_config, ccm_iface->ccm_lib_config.table);
+            ucc_error("lib_init failed for component: %s\n", ccm_iface->name);
+            goto error;
+        }
+        ccm_lib->log_component = ccm_config->log_component;
+        snprintf(ccm_lib->log_component.name, strlen(ccm_iface->ccm_lib_config.prefix),
+                 "%s", ccm_iface->ccm_lib_config.prefix);
+        ccm_lib->priority = (-1 == ccm_config->priority) ?
+            ccm_iface->priority : ccm_config->priority;
+        ucs_config_parser_release_opts(ccm_config, ccm_iface->ccm_lib_config.table);
+        lib->libs[lib->n_libs_opened++] = ccm_lib;
+        ucc_info("lib_prefix \"%s\": initialized component \"%s\" priority %d\n",
+                 config->full_prefix, ccm_iface->name, ccm_lib->priority);
+    }
+    return UCS_OK;
+
+error:
+    if (lib->libs) free(lib->libs);
+    return status;
+}
+
+ucc_status_t ucc_lib_init(const ucc_lib_params_t *params,
+                          const ucc_lib_config_h config,
+                          ucc_lib_h *ucc_lib)
+{
+    ucs_status_t status;
+    ucc_lib_info_t *lib;
+
+    if (ucc_lib_data.n_ccms_loaded == 0) {
+        return UCC_ERR_NO_MESSAGE;
+    }
+
+    lib = malloc(sizeof(*lib));
+    if (!lib) {
+        status = UCC_ERR_NO_MEMORY;
+        goto error;
+    }
+    lib->full_prefix= strdup(config->full_prefix);
+
+    status = ucc_lib_config_components_check(lib, &config->ccms);
+    if (UCS_OK != status) {
+        ucc_error("Unsupported \"UCC_COLL_COMPONENTS\" value\n");
+        goto error;
+    }
+    status = ucc_lib_init_filtered(params, (const ucc_lib_config_t*)config, lib);
+    if (UCS_OK != status) {
+        goto error;
+    }
+
+    if (lib->n_libs_opened == 0) {
+        ucc_error("UCC lib init: no plugins left after filtering by params\n");
+        status = UCC_ERR_NO_MESSAGE;
+        goto error;
+    }
+
+    *ucc_lib = lib;
+    //TODO: move to appropriate place
+    //ucs_config_parser_warn_unused_env_vars_once("UCC_");
+    return UCC_OK;
+
+error:
+    *ucc_lib = NULL;
+    if (lib) free(lib);
+    return status;
+}
+
+ucc_status_t ucc_lib_config_read(const char *env_prefix,
```

Comment:
Why not re-use UCX code for this?

Reply:
We actually exactly re-using UCX code. This "ucc_lib_config_read" is our ucc.h API and its implementation is essentially ucs_config_parser_fill_opts but with minor change: we add "UCC_" prefix.